cuda/components/memory.cuh

```diff
@@ -0,0 +1,789 @@
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CUDA_COMPONENTS_MEMORY_CUH_
+#define GKO_CUDA_COMPONENTS_MEMORY_CUH_
+
+
+#include <type_traits>
+
+
+#include <ginkgo/core/base/math.hpp>
+
+
+#include "cuda/base/types.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace cuda {
+
+
+__device__ __forceinline__ uint32 convert_generic_ptr_to_smem_ptr(void* ptr)
+{
+// see
+// https://github.com/NVIDIA/cutlass/blob/
+//     6fc5008803fe4e81b81a836fcd3a88258f4e5bbf/
+//     include/cutlass/arch/memory_sm75.h#L90
+// for reasoning behind this implementation
+#if (!defined(__clang__) && __CUDACC_VER_MAJOR__ >= 11)
+    return static_cast<uint32>(__cvta_generic_to_shared(ptr));
+#elif (!defined(__clang__) && CUDACC_VER_MAJOR__ == 10 && \
+       __CUDACC_VER_MINOR__ >= 2)
+    return __nvvm_get_smem_pointer(ptr);
+#else
+    uint32 smem_ptr;
+    asm("{{ .reg .u64 smem_ptr; cvta.to.shared.u64 smem_ptr, %1; cvt.u32.u64 "
+        "%0, smem_ptr; }}"
+        : "=r"(smem_ptr)
+        : "l"(ptr));
+    return smem_ptr;
+#endif
+}
+
+
+__device__ __forceinline__ uint32 membar_acq_rel()
+{
+#if __CUDA_ARCH__ < 700
+    asm volatile("membar.gl;" ::: "memory");
+#else
+    asm volatile("fence.acq_rel.gpu;" ::: "memory");
+#endif
+}
+
+
+__device__ __forceinline__ uint32 membar_acq_rel_shared()
+{
+#if __CUDA_ARCH__ < 700
+    asm volatile("membar.cta;" ::: "memory");
+#else
+    asm volatile("fence.acq_rel.cta;" ::: "memory");
+#endif
```

Comment:
pre-Volta doesn't have relaxed/acq/rel memory semantics, see [this example](https://godbolt.org/z/dh8qEY1Tx)

Reply:
My fear was actually that we might run into old CUDA versions that don't support `fence`, but since we officially only support CUDA 10.1+, this is fine.

Reply:
the requirement (fense require sm>=70) is also listed in the end of https://docs.nvidia.com/cuda/archive/9.0/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-membar